cirq/ops/common_channels.py

```diff
@@ -31,109 +31,196 @@
 class AsymmetricDepolarizingChannel(gate_features.SingleQubitGate):
     """A channel that depolarizes asymmetrically along different directions."""
 
-    def __init__(self, p_x: float, p_y: float, p_z: float) -> None:
+    def __init__(self,
+                 p_x: Optional[float] = None,
+                 p_y: Optional[float] = None,
+                 p_z: Optional[float] = None,
+                 num_qubits: int = 1,
+                 error_probabilities: Optional[Dict[str, float]] = None
+                ) -> None:
         r"""The asymmetric depolarizing channel.
 
-        This channel applies one of four disjoint possibilities: nothing (the
-        identity channel) or one of the three pauli gates. The disjoint
-        probabilities of the three gates are p_x, p_y, and p_z and the
-        identity is done with probability 1 - p_x - p_y - p_z. The supplied
-        probabilities must be valid probabilities and the sum p_x + p_y + p_z
-        must be a valid probability or else this constructor will raise a
-        ValueError.
+        This channel applies one of 4**n disjoint possibilities: nothing (the
+        identity channel) or one of the 4**n pauli gates. The disjoint
+        probabilities of the 4**n gates.
 
         This channel evolves a density matrix via
 
             $$
-            \rho \rightarrow (1 - p_x - p_y - p_z) \rho
-                    + p_x X \rho X + p_y Y \rho Y + p_z Z \rho Z
+            \sum_i p_i Pi \rho Pi
             $$
 
+        Where i varies from 0 to 4**n-1, representing the 4**n Pauli gates,
+        represented by Pi. The input \rho is the density matrix before the
+        depolarization.
+
         Args:
             p_x: The probability that a Pauli X and no other gate occurs.
             p_y: The probability that a Pauli Y and no other gate occurs.
             p_z: The probability that a Pauli Z and no other gate occurs.
+            num_qubits: Number of qubits
+            error_probabilities: Dictionary of string (Pauli operator) to its
+                probability
 
         Raises:
             ValueError: if the args or the sum of args are not probabilities.
         """
-
-        self._p_x = value.validate_probability(p_x, 'p_x')
-        self._p_y = value.validate_probability(p_y, 'p_y')
-        self._p_z = value.validate_probability(p_z, 'p_z')
-        self._p_i = 1 - value.validate_probability(p_x + p_y + p_z,
-                                                   'p_x + p_y + p_z')
+        if error_probabilities is not None:
+            for k in error_probabilities.keys():
```

Comment:
Actually, we can remove the `num_qubits` argument and infer the number of qubits from the length of the first key in `error_probabilities`. Sorry for missing this earlier.

(We should still keep the for loop here to ensure all keys are of the same length. We can also keep the internal `_num_qubits` since you're making use of it elsewhere.)